### PR TITLE
Add support for left padding to HookedTransformer's to_tokens method.

### LINKED ISSUE #240: [Proposal] Add support for padding/masking to the attention computation (Minor feature request)
### Proposal 

Models like BERT allow padding to the attention pattern, where the attention pattern is always zero for <PAD> tokens, ie the model cannot look at those tokens. This can also be done by explicitly passing in a Boolean attention mask, for models which do not have a separate PAD and END_OF_TEXT/START_OF_TEXT (EOS/BOS) tokens like GPT-2. 

### Motivation

This is mildly useful to have for even GPT-style models, as it means that if you eg input a batch of prompts of different lengths, they can be padded on the left, so that `logits[:, -1, :]` gets you the final token of each prompt, rather than logits[:, -1, :] being meaningless for shorter prompts.

I particularly want this to be part of exploratory analysis demo, so that when people input individual prompts it's harder to shoot yourself in the foot by having them be of different lengths. 

### Alternatives

The main alternative is just to do padding manually, or massage your prompts to all have the same length. 

### Additional context

This would also need to change absolute positional embeddings to use the attention mask, so that the first real token in each prompt gets the 0th positional embedding (it'll break things otherwise). I would implement this by having an attention mask object that gets optionally created by `to_tokens` if you give it a `left_pad=True` option, or possibly have a `to_tokens_left_pad` option, and which gets passed to `pos_embed` and to `attn`. Relative positional embedding schemes *should* be unaffected. 

This isn't that big a deal, just a mildly nice feature to have.

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/Farama-Foundation/MiniGrid/issues) in the repo (**required**)

### PR DESCRIPTION
**Pull request by JayBaileyCS**
# Description

- Added keyword argument left_pad (default False to avoid breaking anything) to HookedTransformer's to_tokens method. This keyword argument shifts the tokens tensor so that padding is on the left instead of the right.
- Added tests to support this change.
- Added this change to the Exploratory Analysis Demo.
- Fixed a test in test_utils that was failing by enforcing dtype=np.int64
- Some automated formatting changes.

Fixes #240 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

### CODE REVIEW

**Comment by neelnanda-io** [comment]
You'll need to also add support for custom attention masks as in #240 for this to work. Otherwise the model will just attend to the pad tokens! (This is not an issue for right pad because generative models attend backwards only)

File: transformer_lens/HookedTransformer.py
Comment (jbloomAus) [inline transformer_lens/HookedTransformer.py:1]: This seems unnecessary to implement ourselves. I think if we're doing this, we should have it as a setting when you load in the model, set the padding to left-side. 

```
>>> model.tokenizer.padding_side = "left"
>>> prompts = ["Hello", "How are you?", "I am fine, thank you."]
>>> tokens_without_padding = model.to_tokens(prompts, prepend_bos=True, left_pad=False)
>>> tokens_without_padding

tensor([[50256, 50256, 50256, 50256, 50256, 50256, 50256, 15496],
        [50256, 50256, 50256, 50256,  2437,   389,   345,    30],
        [50256,    40,   716,  3734,    11,  5875,   345,    13]])
```

File: tests/unit/test_tokenization_methods.py
Comment (jbloomAus) [inline tests/unit/test_tokenization_methods.py:1]: Thanks for writing tests. 
Would be good to avoid single letter variable names.

File: demos/Exploratory_Analysis_Demo.ipynb
Comment (jbloomAus) [inline demos/Exploratory_Analysis_Demo.ipynb:1]: Last few sentences are unnecessary/confusing. 

"In this case, it's not necessary since all our prompts are of the same length. However, since you get the same result either way when no padding is applied, we include it anyway so you don't have to worry about it."

I think we should just have the warning and a seperate example with different length sentences as an aside.

**Review by jbloomAus — CHANGES_REQUESTED** [review]
Hey Jay, 

Thanks for the PR.

Some things I'd like to address:

- [ ] 1. AFAICT, you haven't fixed the positional embedding which will be wrong if we left pad (see Neel's comments on the original issue).
- [ ] 2. Similarly, we haven't started masking the attention to padded tokens which may could be necessary for coherent output. 
- [ ]  As commented, I don't think we should have our own code for doing the padding, only updating the model's tokenizer. 

Thanks for starting this!

File: transformer_lens/HookedTransformer.py
Comment (JayBaileyCS) [reply transformer_lens/HookedTransformer.py:1]: I'm a bit confused here. My understanding is, the tokenizer (taken from HuggingFace?) already has a padding_side argument that we can use, so we can just use that?

I assume the attention_mask thing will need to be done for the functions where the model is actually being run for inference, since we want the attention_mask to be used to affect attention patterns in the forward pass. 

That said, with tokens_without_padding, why would we be passing in left_pad=False in this scenario if we're not using a left_pad argument? And it seems the tensor is left padded despite left_pad = False here?

Comment (jbloomAus) [reply transformer_lens/HookedTransformer.py:1]: > I'm a bit confused here. My understanding is, the tokenizer (taken from HuggingFace?) already has a padding_side argument that we can use, so we can just use that?

Correct, but it should instantiated with the class since it's used automatically by various methods.

> I assume the attention_mask thing will need to be done for the functions where the model is actually being run for inference, since we want the attention_mask to be used to affect attention patterns in the forward pass.

Yup!

> That said, with tokens_without_padding, why would we be passing in left_pad=False in this scenario if we're not using a left_pad argument? And it seems the tensor is left padded despite left_pad = False here?

This is because I updated the padding side manually. What I did above is not ideal because I'm working around our to_tokens api. I'd rather the api itself make use of the tokenizer method but the actual control flow still make use of the tokenizer property. One solution might be to instantiate the class with a padding side rather than to ever pass it into to_tokens. This way the model class has a consistent padding property.

**Comment by JayBaileyCS** [comment]
This has gotten too big so it's being replaced with https://github.com/neelnanda-io/TransformerLens/pull/323
